content/recipes/authoring/kata-snippets/submission-tests.md

```diff
@@ -0,0 +1,88 @@
+- **Use appropriate random utilities available in your language**. Know how to use RNGs, how to randomly generate types of inputs you need, be it numbers in a range, large numbers, strings, big integers, floating point values, collections.  Know how to do random sampling of collections, how to shuffle them, how to avoid duplicates. See this [obligatory XKCD comic](https://xkcd.com/221/) for how NOT to do random tests.
+- **Random tests shouldn't be grouped by output type or behavior**, especially if the expected output is a boolean variable (e.g. checking that some input satisfies some criteria), or when it comes to error checking (solution throwing an exception in some specific situations). Hence, the order of tested scenarios should be unpredictable too. One possible way to achieve this is to generate and collect a set of random inputs for all required scenarios and shuffle them before the actual testing.
+- Under some rare circumstances, it's allowed to use so called [**randomized tests**](/glossary/#randomized-tests) instead of fully random ones. For some types of problems (for example, kata simulating chess, or problems with small set of possible inputs) it may be too complex or infeasible to generate inputs randomly. In such rare cases, it's accepted to use a predefined, hard-coded set of inputs (and possibly expected outputs too). Before the tests are run, the set of inputs should be shuffled, and eventually randomly sampled at each run to make the hard-coding of the results more tedious for the user. If possible, some additional transformation can be applied randomly (for example array can be reversed, game board can be rotated, sides can be flipped) if it can be easily accounted for in the result of the test.
+- **Random tests should be run after fixed tests.** Not all testing frameworks allow for easy ordering of tests, but fixed tests should run, and eventually fail, before random tests.
+- **Difficulty should be consistent between test runs.** When randomized ranges are very large, it becomes possible for some users to receive many small, easy inputs while other users receive the exact opposite. Make sure that your random tests are built in a way which minimizes the chances for such a situation to occur. If you want to test difficult inputs, split the test cases into a set of easy ones and a set of difficult ones, and test them separately.
+- If it's not possible to build fixed tests in a way they'd be able to catch most of the failures and you have to rely on random tests to catch errors, **make debugging of random tests easy.** Sometimes, when random tests fail, they are very difficult to debug because the input is very large and cannot be easily read. If necessary, split your random tests into two batches: one with small, debuggable inputs, and the other one with proper, large input values. Note that both parts should still contain all applicable scenarios.
```

Comment:
```suggestion
- **Make debugging of random tests easy if you need to rely on them to extensively check the correctness of the user's solution**, if it's not possible to build fixed tests in a way they'd be able to catch most of the holes in the logic of the user. Sometimes, when random tests fail, they are very difficult to debug because the input is very large and cannot be easily read. If necessary, split your random tests into two batches: one with small, debuggable inputs, and the other one with proper, large input values. Note that both parts should still contain all applicable scenarios.
```

Reply:
LGTM